### PR TITLE
[bpf_map_lookup_batch] Fix bpf_map_lookup_batch data retrieval and add test cases for percpu hash map batch lookup

### PR DESCRIPTION
**Pull request by shpalani**
## Description

Problem:
 1. bpf_map_lookup_batch failed to retrieve data for per-cpu map types.
 2. There were no test cases for per-cpu hash map types.

Resolution:
1. Fixed the ebpf_map_lookup_element_batch function to get the right value_size for the per-cpu map types.
2. Added test cases for the per-cpu hash.
  - BPF_MAP_TYPE_PERCPU_HASH
  - BPF_MAP_TYPE_LRU_PERCPU_HASH

## Testing

_Do any existing tests cover this change? Yes.
 Are new tests needed? Yes

## Documentation

_Is there any documentation impact for this change? No

## Installation

_Is there any installer impact for this change? No

### CODE REVIEW

File: tests/unit/libbpf_test.cpp [hunk 1]
```diff
@@ -3388,34 +3393,59 @@ _test_batch_iteration_maps(fd_t& map_fd, uint32_t batch_size, bpf_map_batch_opts
...
+    std::sort(returned_keys.begin(), returned_keys.end());
     for (uint32_t i = 0; i < batch_size; i++) {
         REQUIRE(returned_keys[i] == i);
-        REQUIRE(returned_values[i] == static_cast<uint64_t>(i) * 2ul);
     }
 }
 
 void
-_test_maps_batch(bpf_map_type map_type)
+_test_maps_batch(bpf_map_type map_type, bool percpu)
```
Comment (matthewige) [inline tests/unit/libbpf_test.cpp:3426]: It looks like percpu can be inferred from the map_type. Is that correct? If so, perhaps it need not be an input parameter and can just be determined based on the map type.

File: libs/api/ebpf_api.cpp [hunk 2]
```diff
@@ -778,9 +782,8 @@ _update_map_element_batch(
                 uint8_t* source_value = (uint8_t*)value + (key_index + index) * value_size;
                 uint8_t* destination_key = request->data + index * (key_size + value_size);
                 uint8_t* destination_value = request->data + index * (key_size + value_size) + key_size;
-                if (key_size > 0) {
-                    std::copy(source_key, source_key + key_size, destination_key);
-                }
+
+                std::copy(source_key, source_key + key_size, destination_key);
```
Comment (matthewige) [inline libs/api/ebpf_api.cpp:786]: Can you explain why the if check was removed here?

File: tests/unit/libbpf_test.cpp [hunk 3]
```diff
@@ -3361,21 +3361,26 @@ TEST_CASE("bind_tail_call_max_exceed", "[libbpf]")
...
     uint32_t* in_batch = nullptr;
     uint32_t out_batch = 0;
     std::vector<uint32_t> returned_keys(0);
     std::vector<uint64_t> returned_values(0);
-    uint32_t requested_batch_size = (batch_size / batch_size_divisor) - 2;
+    int32_t requested_batch_size = (batch_size / batch_size_divisor) - 2;
     uint32_t batch_size_count = 0;
     int result = 0;
 
+    if (requested_batch_size <= 0) {
```
Comment (matthewige) [inline tests/unit/libbpf_test.cpp:3377]: Can you explain why this change was needed? Are we hitting this now because of changes in this PR?

File: libs/api/ebpf_api.cpp (hunk 2 above)
Comment (shpalani) [reply libs/api/ebpf_api.cpp:786]: On line 756, the key_size is already being checked.
    if (key_size == 0 || value_size == 0 || input_count == 0) {
        result = EBPF_INVALID_ARGUMENT;
        goto Exit;
    }

File: tests/unit/libbpf_test.cpp (hunk 1 above)
Comment (matthewige) [reply tests/unit/libbpf_test.cpp:3426]: Re-opening this. The other code change in your PR uses BPF_MAP_TYPE_PER_CPU() macro. Is there a reason this cannot be used?

**Review by matthewige — APPROVED** [review]

File: tests/unit/libbpf_test.cpp [hunk 4]
```diff
@@ -3388,26 +3393,32 @@ _test_batch_iteration_maps(fd_t& map_fd, uint32_t batch_size, bpf_map_batch_opts
...
 
         in_batch = &out_batch;
     }
 
     REQUIRE(returned_keys.size() == batch_size);
-    REQUIRE(returned_values.size() == batch_size);
-
-    std::sort(returned_keys.begin(), returned_keys.end());
-    std::sort(returned_values.begin(), returned_values.end());
+    REQUIRE(returned_values.size() == batch_size * value_size);
```
Comment (saxena-anurag) [inline tests/unit/libbpf_test.cpp:3407]: nit: should it be `batch_size_count` instead of `batch_size`?

Comment (saxena-anurag) [reply tests/unit/libbpf_test.cpp:3407]: here and also in the following code.

Comment (shpalani) [reply tests/unit/libbpf_test.cpp:3407]: >> should it be batch_size_count instead of batch_size?

No, it should be batch_size only, which is the total requested.

FYI, batch_size_count is used to retrieve requested_batch_size, where small blocks of key/value are retrieved in iteration.

File: tests/unit/libbpf_test.cpp [hunk 5]
```diff
@@ -3429,10 +3459,13 @@ _test_maps_batch(bpf_map_type map_type)
 
     uint32_t batch_size = 20000;
     std::vector<uint32_t> keys(batch_size);
-    std::vector<uint64_t> values(batch_size);
+    std::vector<uint64_t> values(batch_size * value_size);
```
Comment (shankarseal) [inline tests/unit/libbpf_test.cpp:3462]: consider
```
using per_cpu_map_value = vector<uint64_t>;
using batch_of_values = vector<per_cpu_map_values>;

batch_of_values values(batch_count, per_cpu_map_value(num_cpu_count));

for (i = 0; i < batch_count; i++) {
   for (cpu = 0; cpu < num_cput_count; cpu++) {
      values[i][cpu] = 2*i;
   }
}
````

Comment (shpalani) [reply tests/unit/libbpf_test.cpp:3462]: The above snippet is a vector of vector. 
It is better to keep it simple as a vector (instead of vector of vector), where it suffices for both non-per-cpu (num of CPU is 1) and per-cpu.

**Review by Alan-Jowett — APPROVED** [review]

File: tests/unit/libbpf_test.cpp (hunk 5 above)
Comment (shpalani) [reply tests/unit/libbpf_test.cpp:3462]: As discussed offline with you regarding the comment of considering the approach of vector of vector, the buffer/values need to be flattened again to provide to ebpf update batch API.
